ui-spacetimechart/src/hooks/useMouse.ts

```diff
@@ -0,0 +1,188 @@
+        ...state,
+        down: { ctrl: e.ctrlKey, shift: e.shiftKey },
+      }));
+    },
+    [dom]
+  );
+  const upHandler = useCallback(() => {
+    if (!dom) return;
+
+    setMouseState((state) => ({ ...state, down: undefined }));
+  }, [dom]);
+  const moveHandler = useCallback(
+    (event: MouseEvent) => {
+      if (!dom) return;
+
+      const position = getEventPosition(event, dom);
+      const { x, y } = position;
+      const width = dom.offsetWidth;
+      const height = dom.offsetHeight;
+      const isHover = x >= 0 && x <= width && y >= 0 && y <= height;
+      setMouseState((state) => ({
+        ...state,
+        position,
+        isHover,
+      }));
+
+      const { onMouseMove } = handlersRef.current;
+      if (onMouseMove)
+        onMouseMove({
+          event,
+          isHover,
+          position,
+          // TODO
+          data: { position: NaN, time: NaN },
+        });
+    },
+    [dom]
+  );
+  const wheelHandler = useCallback(
+    (event: WheelEvent) => {
+      const { onZoom } = handlersRef.current;
+      if (onZoom && dom)
+        onZoom({
+          delta: getEventWheelDelta(event),
+          position: getEventPosition(event, dom),
+          event,
+        });
+    },
+    [dom]
+  );
+
+  // Bind event handlers:
+  useEffect(() => {
+    if (!dom) return;
+    dom.addEventListener('click', clickHandler);
+    return () => {
+      dom.removeEventListener('click', clickHandler);
+    };
+  }, [dom, clickHandler]);
+  useEffect(() => {
+    if (!dom) return;
+    dom.addEventListener('mousedown', downHandler);
+    return () => {
+      dom.removeEventListener('mousedown', downHandler);
+    };
+  }, [dom, downHandler]);
+  useEffect(() => {
+    document.addEventListener('mouseup', upHandler);
+    return () => {
+      document.removeEventListener('mouseup', upHandler);
+    };
+  }, [upHandler]);
+  useEffect(() => {
+    document.addEventListener('mousemove', moveHandler);
+    return () => {
+      document.removeEventListener('mousemove', moveHandler);
+    };
+  }, [moveHandler]);
+  useEffect(() => {
+    if (!dom) return;
+    dom.addEventListener('wheel', wheelHandler);
+    return () => {
+      dom.removeEventListener('wheel', wheelHandler);
+    };
+  }, [dom, wheelHandler]);
+
+  // Listen to "up" and "down" updates:
+  useEffect(() => {
+    const { onPan } = handlersRef.current;
+
+    if (down) {
+      // Start panning:
+      setPanningState({
+        type: 'panning',
+        initialPosition: position,
+        initialData: getData(position),
+      });
+    } else {
+      // Stop panning:
+      if (panningState.type === 'panning' && onPan)
+        onPan({
+          isPanning: false,
+          position,
+          initialPosition: panningState.initialPosition,
+          data: getData(position),
+          initialData: panningState.initialData,
+        });
+
+      if (panningState.type !== 'idle') setPanningState({ type: 'idle' });
+    }
```

Comment:
Can we simplify the function like this ? 
-> if onPan is not defined, then we do nothing

```suggestion
    if (onPan) {
      if (down) {
        // Start panning:
        setPanningState({
          type: 'panning',
          initialPosition: position,
          initialData: getData(position),
        });
      } else if (panningState.type === 'panning') {
        // Stop panning:
        onPan({
          isPanning: false,
          position,
          initialPosition: panningState.initialPosition,
          data: getData(position),
          initialData: panningState.initialData,
        });
        setPanningState({ type: 'idle' });
      }
```

Reply:
I fear weird bugs with you code in some very edge cases, such that stopping giving an `onPan` prop while panning, and giving it back later. That's why I preferred to always keep the internal state updated, even when there is no `onPan` prop given.

Reply:
OK got it :+1: